Ensure bundle not created on fork block

liuchengxu opened the pull request:
This PR primarily adds a check against the bundle header to reject the bundles that are created on a primary chain fork. The second commit is trivial to address a spec comment.

Close #1195

### Code contributor checklist:
* [x] I have read, understood and followed [contributing guide](https://github.com/subspace/subspace/blob/main/CONTRIBUTING.md)

liuchengxu committed:
Check if bundle was created on a valid primary block

A valid primary block means it's on the same branch as the block
the runtime corresponding to, reject the bundle if it's created from a primary chain fork.

liuchengxu committed:
Rename `withdraw_decreased_stake` to `withdraw_stake`

Resolves https://www.notion.so/Namespaced-Execution-Rough-Spec-e9918e6d17aa4b24854d7eb150f93fcc?d=1c8847cb7e56478ead280e9167ad963b#1e2ac56cc50d44b99d345e1a08dc18e2

NingLin-P commented on crates/pallet-domains/src/lib.rs:
```diff
@@ -584,6 +586,29 @@ impl<T: Config> Pallet<T> {
+            match pallet_receipts::PrimaryBlockHash::<T>::get(
+                DomainId::SYSTEM,
+                bundle.header.primary_number,
+            ) {
```
Using the system domain's `PrimaryBlockHash` for the core domain may be inappropriate. For example, if a core domain is running slow or stalls and the corresponding `PrimaryBlockHash` of the system domain may have already been pruned, thus causing this check always fails for this core domain.

NingLin-P replied:
Or perhaps we should skip further checking if `PrimaryBlockHash::<T>::get` return `None` and it is a core bundle, to make this check a best effort for the core domain (it should cover most cases).

liuchengxu replied:
You're right, the check in pallet-domains is only suitable for the system domain bundle. But we need to find another way for the core domain bundles, skipping it for the core domain bundles does not solve the problem, we want it to be guaranteed, let me think more about this.

liuchengxu replied:
Changed the corresponding check to system bundle only and added TODO for the core domains as it's not feasible right now.

liuchengxu committed:
pallet-domains: only check of bundle not created on a primary fork for the system bundle

Within pallet-domain, only the system domain receipts are maintained in
pallet-receipts, therefore we can only check the system domain bundle.

NingLin-P commented on crates/pallet-domains/src/lib.rs:
```diff
@@ -584,6 +586,29 @@ impl<T: Config> Pallet<T> {
+        if !bundle_created_on_valid_primary_block {
+            log::error!(
+                target: "runtime::domains",
+                "Bundle is probabaly created on a primary fork #{:?}, expected: {:?}, got: {:?}",
```
```suggestion
                "Bundle is probably created on a primary fork #{:?}, expected: {:?}, got: {:?}",
```

It is also possible that the bundle points to a future block that this node doesn't sync yet.

liuchengxu replied:
It's indeed possible, but unlike the receipts, we probably don't want this optimization as we'd like the fork-aware bundles to be strongly ensured.

liuchengxu committed:
pallet-domain-registry: check of bundle not created on a primary fork for the core domain bundles

liuchengxu committed:
Add TODO of core domain bundles created on primary forks

liuchengxu committed:
Merge branch 'main' of https://github.com/subspace/subspace into ensure-bundle-not-created-on-fork-block

liuchengxu committed:
Merge branch 'main' into ensure-bundle-not-created-on-fork-block
